desc/magnetic_fields.py

```diff
@@ -795,3 +831,464 @@ def odefun(rpz, s):
     r = x[:, :, 0].T.reshape((len(phis), *rshape))
     z = x[:, :, 2].T.reshape((len(phis), *rshape))
     return r, z
+
+
+class CurrentPotentialField(MagneticField, FourierRZToroidalSurface):
+    """Magnetic field due to a surface current potential on a toroidal surface.
+
+        surface current K is assumed given by
+         K = ∇ Φ x n
+        where n is the winding surface unit normal
+              Phi is the current potential function,
+              which is a function of theta and zeta
+        this function then uses biot-savart to find the
+        B field from this current density K on the surface
+
+        Note: potential function derivatives, if AD is used,
+        will not be vectorized over the params
+
+    Parameters
+    ----------
+    potential : callable
+        function to compute the current potential. Should have a signature of
+        the form potential(theta,zeta,**params) -> ndarray.
+        theta,zeta are poloidal and toroidal angles on the surface
+    surface_grid : Grid,
+        grid upon which to evaluate the surface current density K
+    params : dict, optional
+        default parameters to pass to potential function (and its derivatives)
+    potential_dtheta: callable
+        function to compute the theta derivative of the current potential
+        if None, will use AD to calculate
+    potential_dzeta: callable
+        function to compute the theta derivative of the current potential
+        if None, will use AD to calculate
+    R_lmn, Z_lmn : array-like, shape(k,)
+        Fourier coefficients for winding surface R and Z in cylindrical coordinates
+    modes_R : array-like, shape(k,2)
+        poloidal and toroidal mode numbers [m,n] for R_lmn.
+    modes_Z : array-like, shape(k,2)
+        mode numbers associated with Z_lmn, defaults to modes_R
+    NFP : int
+        number of field periods
+    sym : bool
+        whether to enforce stellarator symmetry. Default is "auto" which enforces if
+        modes are symmetric. If True, non-symmetric modes will be truncated.
+    rho : float [0,1]
+        flux surface label for the toroidal surface
+    name : str
+        name for this surface
+    check_orientation : bool
+        ensure that this surface has a right handed orientation. Do not set to False
+        unless you are sure the parameterization you have given is right handed
+        (ie, e_theta x e_zeta points outward from the surface).
+    surface: FourierRZToroidalSurface, optional, default None
+        Existing FourierRZToroidalSurface object to create a
+        CurrentPotentialField with, if provided will use this
+        object's R_lmn, Z_lmn etc instead of any passed-in values
+
+    """
+
+    _io_attrs_ = (
+        MagneticField._io_attrs_
+        + FourierRZToroidalSurface._io_attrs_
+        + [
+            "_surface_grid",
+            "_params",
+        ]
+    )
+
+    def __init__(
+        self,
+        potential,
+        surface_grid,
```

Comment:
could default to None and have preset default based on resolution of surface? Or remove entirely and just make it defined within the compute method where its needed?

Reply:
It is hard to think of a good default resolution, as even with simple potentials and simple surfaces (like a rotating ellipse), you still need pretty high surface grid resolution to accurately resolve the magnetic fields inside (I have found at least)

Reply:
we can